File: vm/vm.c

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define STACK_SIZE 1024
#define STRING_BUFFER_SIZE 1024

#define TYPE_NUMBER 0x1111000000000001
#define TYPE_STRING 0x1111000000000002
#define TYPE_COMMAND 0xCCCC000000000001

#define ID_RETURN 0
#define ID_PUTS 1
#define ID_OPEN_WINDOW 12

typedef struct {
    int64_t stack[STACK_SIZE];
    int64_t *top;
    int64_t *end;
    char string_buffer[STRING_BUFFER_SIZE];
    size_t string_offset;
} Stack;

__declspec(dllexport) void init_stack(Stack *stack) {
    stack->top = stack->stack;
    stack->end = stack->stack + STACK_SIZE;
    stack->string_offset = 0;
}

__declspec(dllexport) void push_number(Stack *stack, int64_t value) {
    if (stack->top + 2 > stack->end) {
        fprintf(stderr, "Stack overflow: Not enough space to push number.\n");
        exit(EXIT_FAILURE);
    }
    *stack->top++ = value;
    *stack->top++ = TYPE_NUMBER;
}

__declspec(dllexport) int64_t pop_number(Stack *stack) {
    if (stack->top - 2 < stack->stack) {
        fprintf(stderr, "Stack underflow: No numbers to pop.\n");
        exit(EXIT_FAILURE);
    }
    if (*--stack->top != TYPE_NUMBER) {
        fprintf(stderr, "Type error: Expected number type.\n");
        exit(EXIT_FAILURE);
    }
    return *--stack->top;
}

__declspec(dllexport) void push_string(Stack *stack, const char *string) {
    size_t len = strlen(string);
    size_t aligned_len = (len + 7) & ~7;

    if (stack->string_offset + aligned_len > STRING_BUFFER_SIZE) {
        fprintf(stderr, "String buffer overflow: Not enough space to push string.\n");
        exit(EXIT_FAILURE);
    }

    char *dest = stack->string_buffer + stack->string_offset;
    memcpy(dest, string, len);
    memset(dest + len, 0, aligned_len - len);

    size_t offset = stack->string_offset;
    stack->string_offset += aligned_len;

    if (stack->top + 3 > stack->end) {
        fprintf(stderr, "Stack overflow: Not enough space to push string metadata.\n");
        exit(EXIT_FAILURE);
    }
    *stack->top++ = offset;
    *stack->top++ = len;
    *stack->top++ = TYPE_STRING;
}

__declspec(dllexport) const char *pop_string(Stack *stack) {
    if (stack->top - 3 < stack->stack) {
        fprintf(stderr, "Stack underflow: No strings to pop.\n");
        exit(EXIT_FAILURE);
    }
    if (*--stack->top != TYPE_STRING) {
        fprintf(stderr, "Type error: Expected string type.\n");
        exit(EXIT_FAILURE);
    }
    size_t len = *--stack->top;
    size_t offset = *--stack->top;

    if (offset + len > STRING_BUFFER_SIZE) {
        fprintf(stderr, "String buffer corruption detected.\n");
        exit(EXIT_FAILURE);
    }
    return stack->string_buffer + offset;
}

__declspec(dllexport) void execute_bytecode(Stack *stack, const int64_t *bytecode, size_t bytecode_len) {
    size_t i = 0;

    while (i < bytecode_len) {
        int64_t op = bytecode[i++];

        if (op == TYPE_NUMBER) {
            int64_t value = bytecode[i++];
            push_number(stack, value);
        } else if (op == TYPE_STRING) {
            size_t packed_len = bytecode[i++];
            char buffer[256] = {0};
            for (size_t j = 0; j < packed_len; ++j) {
                int64_t chunk = bytecode[i++];
                memcpy(buffer + j * 8, &chunk, 8);
            }
            push_string(stack, buffer);
        } else if (op == TYPE_COMMAND) {
            int64_t cmd_id = bytecode[i++];
            int64_t num_args = bytecode[i++];

            if (cmd_id == ID_RETURN) {
                printf("Return command executed.\n");
                break;
            } else if (cmd_id == ID_PUTS) {
                const char *s = pop_string(stack);
                printf("%s\n", s);
            } else if (cmd_id == ID_OPEN_WINDOW) {
                int64_t height = pop_number(stack);
                int64_t width = pop_number(stack);
                printf("Open window: width=%lld, height=%lld\n", width, height);
            } else {
                fprintf(stderr, "Unknown command: %lld\n", cmd_id);
                exit(EXIT_FAILURE);
            }
        } else {
            fprintf(stderr, "Unknown operation: %lld\n", op);
            exit(EXIT_FAILURE);
        }
    }
}

File: vm/vm.py
import math

print('hello')

msg = 'hello world'

type_number  = 0x1111_0000_0000_0001
type_string  = 0x1111_0000_0000_0002
type_command = 0xcccc_0000_0000_0001

id_return = 0
id_puts = 1
id_open_window = 12

def packed_string(s):
    data = s.encode('utf8')
    alignment = 8  # 64-bit alignment
    padding_size = (alignment - (len(data) % alignment)) % alignment
    aligned_data = data + b'\x00' * padding_size
    return [int.from_bytes(aligned_data[i:i + 8], byteorder='big', signed=False) for i in range(0, len(aligned_data), 8)]


def print_stack(stack):
    print(f"Stack contents: ({len(stack)})")
    i = len(stack) - 1  # Start from the back since type is on top
    while i >= 0:
        if stack[i] == type_number:
            print(f"Number: {stack[i - 1]}")
            i -= 2
        elif stack[i] == type_string:
            length = stack[i - 1]
            string_ints = stack[i - 2 - length:i - 2]
            string_value = b''.join(int.to_bytes(num, 8, byteorder='big', signed=False) for num in string_ints).rstrip(b'\0').decode('utf8')
            print(f"String: '{string_value}' (Length: {length*8})")
            i -= 2 + length
        else:
            print(f"Unknown Type Code: {stack[i]}")
            break
    if i != -1:  # If we didn't process the entire stack, we have an issue
        print(f"Stack integrity compromised: {i} items left unprocessed")

def ints_to_bytes(list_of_ints):
    data = b''.join(item.to_bytes(8, byteorder='big', signed=False) for item in list_of_ints)
    return data

def hex_dump(data):
    print("-" * 50)
    for offset in range(0, len(data), 16):
        row = data[offset:offset + 16]
        hex_values = " ".join(f"{byte:02x}" for byte in row)
        ascii_values = "".join(chr(byte) if 32 <= byte <= 126 else "." for byte in row)
        print(f"{offset:08x}  {hex_values:<47}  {ascii_values}")
    print("-" * 50)

def push_number(stack, value):
    stack.extend([value, type_number])

def pop_number(stack):
    if len(stack) < 2 or stack[-1] != type_number:
        raise ValueError("Expected a number on stack")
    stack.pop()  # Remove the type
    return stack.pop()  # Remove and return the value

def push_string(stack, string):
    encoded_string = string.encode("utf-8")
    string_length = len(encoded_string)
    aligned_string_length = math.ceil(string_length / 8) * 8
    padded_string = encoded_string.ljust(aligned_string_length, b'\0')
    string_ints = [int.from_bytes(padded_string[i:i+8], byteorder='big', signed=False) for i in range(0, aligned_string_length, 8)]
    stack.extend([*string_ints, len(string_ints), type_string])

def pop_string(stack):
    if len(stack) < 3 or stack[-1] != type_string:
        raise ValueError("Expected a string on stack")
    stack.pop()  # Remove the type
    length = stack.pop()  # Pop length
    string_ints = [stack.pop() for _ in range(length)][::-1]  # Pop string data
    string_value = b''.join(int.to_bytes(num, 8, byteorder='big', signed=False) for num in string_ints).rstrip(b'\0').decode('utf8')
    return string_value

def call_command(stack, cmd_id):
    #print(f'call command {cmd_id=}')
    if cmd_id == id_open_window:
        h = pop_number(stack)
        w = pop_number(stack)
        print(f'open_window {w=}, {h=}')
    elif cmd_id == id_puts:
        s = pop_string(stack)
        print(s)
    else:
        assert False, f'invalid cmd {cmd_id}'

packed_msg = packed_string(msg)
#print(f'{packed_msg=}')

byte_code = [
    type_string, len(packed_msg), *packed_msg,
    type_command, id_puts,
    type_number, 640,
    type_number, 480,
    type_command, id_open_window,
    type_command, id_return,
]

print('byte_code=')
hex_dump(ints_to_bytes(byte_code))
stack = []  # Use a list as our stack

#print_stack(stack)

# instruction pointer
i = 0

while i < len(byte_code):
    #print()
    op = byte_code[i]
    i += 1
    #print_stack(stack)
    #hex_dump(stack)

    if op == type_number:
        n = byte_code[i]
        i += 1
        push_number(stack, n)

    elif op == type_string:
        packed_len = byte_code[i]
        i += 1
        string_ints = byte_code[i:i + packed_len]
        string_value = b''.join(int.to_bytes(j, 8, byteorder='big', signed=False) for j in string_ints).rstrip(b'\0').decode('utf8')
        i += packed_len
        push_string(stack, string_value)

    elif op == type_command:
        cmd_id = byte_code[i]
        assert cmd_id >= 0
        i += 1

        if cmd_id == id_return:
            print('return')
            break

        call_command(stack, cmd_id)

    else:
        assert False, f'unknown op: {op}'

if stack != []:
    print_stack(stack)
    hex_dump(ints_to_bytes(stack))

